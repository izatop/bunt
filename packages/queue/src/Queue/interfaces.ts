import {Promisify} from "@typesafeunit/util";
import {MessageAbstract} from "./MessageAbstract";

export interface ITransaction<M extends Message> {
    message: M;

    commit(): Promise<HandleResult<M>>;

    rollback(reason?: Error): Promise<IHandleFail<M>>;
}

export interface IChannel<M extends MessageAbstract<any>> {
    readonly type: MessageCtor<M>;
    readonly subscribed: boolean;
    readonly key: string;

    subscribe(): AsyncGenerator<ITransaction<M>>;

    unsubscribe(): void;
}

export interface ITransport {
    send<M extends Message>(message: M): Promisify<void>;

    channel<M extends Message>(type: MessageCtor<M>): IChannel<M>;
}

export interface IMessageHandler<M extends MessageAbstract<any>> {
    (message: M): unknown;
}

export type QueueKeys<T> = Extract<keyof T, string>;

export interface IHandleState<M extends Message> {
    error?: Error;
    status: boolean;
    message: M;
    finishAt: Date;
    runAt: Date;
}

export interface IHandleSuccess<M extends Message> extends IHandleState<M> {
    status: true;
    error?: never;
}

export interface IHandleFail<M extends Message> extends IHandleState<M> {
    status: false;
    error: Error;
}

export type HandleResult<M extends Message> = IHandleSuccess<M> | IHandleFail<M>;

export interface IHandleResultFactory<M extends Message> {
    (): IHandleSuccess<M>;

    (error: Error): IHandleFail<M>;
}

export interface ISubscription<M extends Message> {
    unsubscribe(): Promise<void>;

    listenResult(fn: (result: HandleResult<M>) => unknown): void;
}

export type MessagePayload<T extends MessageAbstract<any>> = T extends MessageAbstract<infer P> ? P : never;

export interface MessageCtor<M extends MessageAbstract<any>> {
    prototype: M;

    readonly channel: string;

    new(message: MessagePayload<M>): M;
}

export type Message<T extends MessageAbstract<any> = MessageAbstract<any>> = T;
