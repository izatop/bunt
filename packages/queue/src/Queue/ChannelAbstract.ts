import {IChannel, ITransaction, Message, MessageCtor} from "./interfaces";

export abstract class ChannelAbstract<M extends Message> implements IChannel<M> {
    public abstract readonly subscribed: boolean;

    public readonly type: MessageCtor<M>;

    constructor(type: MessageCtor<M>) {
        this.type = type;
    }

    public get key(): string {
        return this.type.channel;
    }

    public abstract subscribe(): AsyncGenerator<ITransaction<M>>;

    public abstract unsubscribe(): void;
}
