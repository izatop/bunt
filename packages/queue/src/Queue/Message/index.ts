export * from "./MessageAbstract";
export * from "./TransactionAbstract";
export * from "./TaskAbstract";
