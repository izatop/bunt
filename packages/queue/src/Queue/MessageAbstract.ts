export class MessageAbstract<T> {
    public id?: string;

    public readonly payload: T;

    constructor(payload: T) {
        this.payload = payload;
    }

    public get channel(): string {
        return this.id ?? this.constructor.name;
    }

    public static get channel(): string {
        return this.prototype.channel;
    }
}
