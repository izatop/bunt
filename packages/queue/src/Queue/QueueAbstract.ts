import {IMessageHandler, ITransport, Message, MessageCtor} from "./interfaces";
import {Subscription} from "./Subscription";

export class QueueAbstract<Q extends ITransport> {
    readonly #transport: Q;

    constructor(transport: Q) {
        this.#transport = transport;
    }

    public get transport(): Q {
        return this.#transport;
    }

    public async emit<M extends Message>(message: M): Promise<void> {
        await this.#transport.send(message);
    }

    public subscribe<M extends Message>(type: MessageCtor<M>,
                                        handler: IMessageHandler<M>): Subscription<M> {
        return new Subscription<M>(this.#transport.channel(type), handler);
    }
}
