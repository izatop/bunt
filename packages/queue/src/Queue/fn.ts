import {assert} from "@typesafeunit/util";
import * as crypto from "crypto";
import {IHandleResultFactory, Message} from "./interfaces";

const serializeRe = /^[0-9a-f]{8}:.+$/;

export function serialize<M extends Message>(message: M): string {
    const body = JSON.stringify(message.payload);
    const signature = crypto.createHash("sha1")
        .update(body)
        .digest("hex")
        .substr(0, 8);

    return `${signature}:${body}`;
}

export function unserialize<T = unknown>(message: string): T {
    assert(serializeRe.test(message), "Wrong message format");

    const body = message.substr(9);
    const signature = message.substr(0, 8);
    const compareSignature = crypto.createHash("sha1")
        .update(body)
        .digest("hex")
        .substr(0, 8);

    assert(signature === compareSignature, "Wrong checksum");
    return JSON.parse(body);
}

export function createHandleState<M extends Message>(message: M): IHandleResultFactory<M> {
    const runAt = new Date();
    return ((error?: Error) => {
        if (error) {
            return {runAt, error, message, status: false, finishAt: new Date()};
        }

        return {
            runAt,
            message,
            finishAt: new Date(),
            status: true,
        };
    }) as IHandleResultFactory<M>;
}
