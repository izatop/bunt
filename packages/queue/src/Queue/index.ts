export * from "./fn";
export * from "./interfaces";
export * from "./QueueAbstract";
export * from "./ChannelAbstract";
export * from "./TransactionAbstract";
export * from "./MessageAbstract";
export * from "./Subscription";
export * from "./SimpleQueue";
