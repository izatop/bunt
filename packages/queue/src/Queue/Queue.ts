import {ITransport} from "../interfaces";
import {QueueAbstract} from "./QueueAbstract";

export class Queue<Q extends ITransport> extends QueueAbstract<Q> {

}
