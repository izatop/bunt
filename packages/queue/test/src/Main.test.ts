import {BarMessage} from "./Queue/BarMessage";
import {FooMessage} from "./Queue/FooMessage";
import {TestQueue} from "./Queue/TestQueue";
import {TestTransport} from "./Queue/TestTransport";

describe("Response", () => {
    test("Main", async () => {
        const test = new TestQueue(new TestTransport());
        const queue: any[] = [];
        const rejected: any[] = [];
        const push = (message: any) => queue.push(message);
        const reject = () => {
            throw Error("Rejected");
        };

        await test.emit(new FooMessage(true));
        await test.emit(new BarMessage(1));

        const s1 = test.subscribe(FooMessage, push);
        await s1.unsubscribe();

        const s2 = test.subscribe(BarMessage, reject);
        s2.listenResult(({status, message, error}) => {
            rejected.push({status, message, error});
        });

        await s2.unsubscribe();
        await test.emit(new BarMessage(2));

        const s3 = test.subscribe(BarMessage, push);
        await s3.unsubscribe();

        expect(queue).toMatchSnapshot();
        expect(rejected).toMatchSnapshot();
    });
});
