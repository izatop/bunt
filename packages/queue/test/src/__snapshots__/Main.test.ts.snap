// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response Main 1`] = `
Array [
  FooMessage {
    "payload": true,
  },
  BarMessage {
    "payload": 2,
  },
]
`;

exports[`Response Main 2`] = `
Array [
  Object {
    "error": [Error: Rejected],
    "message": BarMessage {
      "payload": 1,
    },
    "status": false,
  },
]
`;
